### ISSUE TITLE
MNIST TFX Pipeline

### ISSUE THREAD

**Issue by Iso-oss** [tensorflow/tfx#321]
Hello! 

First, I would like to thank you for your work. I am currently trying to create a TFX pipeline capable to train a deep learning model on MNIST data set. ImportExampleGen, StatisticGen and SchemaGen are pretty straight forward to setup with the documentation. My data are stored as tf records and images are put into tf records as bytes. I am now struggling to setup up a Transform components that will reshape images and apply some preprocessings (zero mean, unit standard deviation for instance). How can I use Tensorflow Transform to achieve this step ? 

Thank you !

**Comment by ruoyu90** [comment]
Hi @Iso-oss , thanks for trying out TFX :)

In transform stage, we want to use tensor operations to transform input tensors and a graph representing the transformation will be created and passed down to downstream component. You can refer to [TFT site](https://www.tensorflow.org/tfx/transform/get_started) for more information.

For you specific case, maybe [tf.image](https://www.tensorflow.org/api_docs/python/tf/image) API can help, especially [this](https://www.tensorflow.org/api_docs/python/tf/image/per_image_standardization) one.

**Comment by Iso-oss** [comment]
Thank you for your response ! With your answer I managed to transform images as I wanted. I just have one last issue about shapes. During the reshaping steps if I try to recover values for height and width from inputs with the following code : 

`height = _fill_in_missing(inputs["height"])` (same _fill_in_missing function from the taxi example)
`height = tf.cast(height, tf.int32)`
`height = tf.reshape(height, (1,))`

and the same from width, I get the following error : 

`ValueError : Tensor("map/TensorArrayStack/TensorArrayGatherV3:0", shape=(?, ?, ?, ?), dtype=float32) has invalid shape (?, ?, ?, ?) for FixedLenFeatures: apart from the batch dimension, all dimensions must have known size`

Is there a workaround avoiding hardcoding or passing those by config file ? 

Thank you.

**Comment by ruoyu90** [comment]
IIUC the question is about to get the shape of the input tensor instead of hardcoding the output shape when calling `tf.reshape`? You can always get the shape of the tensor by calling `tf.shape` which will give you the shape of the tensor in the format of List[int]. Note that the first element of the result is the batch size.

**Comment by Iso-oss** [comment]
Thank you, I managed to do the whole pipeline as I wanted. Great work, keep it up !

**Comment by ruoyu90** [comment]
@Iso-oss It seems TF is complaining about mixing graph mode and eager mode. I assume the trainer runs fine? Could you attach the trainer code so that we have better context?

@mdreves could you shed some light on this?

**Comment by mdreves** [comment]
Can you try this with TF version 1.14.

**Comment by Iso-oss** [comment]
The pipeline works now :). I mixed up labels keys. Thank you for your help and keep on with TFX :)

**Comment by zhitaoli** [comment]
Hi @Iso-oss, good to hear that things eventually work. Are you willing to publish your example pipeline, either in this repo or in your own? We are looking to expand example coverage of TFX and some of our engineers are looking at creating an image based example, but we are definitely supportive if anyone in the community wants to contribute.

Thanks in advance!

**Comment by chanshah** [comment]
> The pipeline works now :). I mixed up labels keys. Thank you for your help and keep on with TFX :)

Hi @Iso-oss I'm glad the pipeline is working for you now. Would you be willing to publish / contribute the sample for this pipeline for more developers to benefit from this?

**Comment by gilbblig** [comment]
> Hi @Iso-oss, good to hear that things eventually work. Are you willing to publish your example pipeline, either in this repo (through a PR) or in your own? We are looking to expand example coverage of TFX and some of our engineers are looking at creating an image based example, but we are definitely supportive if anyone in the community wants to contribute.
> 
> Thanks in advance!

Hi @zhitaoli , I'm also looking to apply a pipeline based on TFX and Kubeflow for image classification. Since I just started by replicating a csv-based pipeline, I wonder if there is already some code you'd be willing to share.

**Comment by valeriano-manassero** [comment]
@gilbblig Maybe this can be useful. I just started a collection of on premise Kubeflow TFX pipelines:
https://github.com/valeriano-manassero/tfx-kubeflow-pipelines

There's also a Cifar-10 dataset pipeline.
